services/repository/files/update.go

```diff
@@ -454,35 +535,51 @@ func CreateOrUpdateRepoFile(ctx context.Context, repo *repo_model.Repository, do
 		return nil, err
 	}
 
-	file, err := GetFileResponseFromCommit(ctx, repo, commit, opts.NewBranch, treePath)
-	if err != nil {
-		return nil, err
+	files := []*structs.FileResponse{}
+
+	for _, file := range opts.Files {
+		fileResponse, err := GetFileResponseFromCommit(ctx, repo, commit, opts.NewBranch, filesOptions[file.TreePath].treePath)
+		if err != nil {
+			return nil, err
+		}
+		files = append(files, fileResponse)
 	}
 
 	if repo.IsEmpty {
 		_ = repo_model.UpdateRepositoryCols(ctx, &repo_model.Repository{ID: repo.ID, IsEmpty: false}, "is_empty")
 	}
 
-	return file, nil
+	return files, nil
 }
 
 // VerifyBranchProtection verify the branch protection for modifying the given treePath on the given branch
-func VerifyBranchProtection(ctx context.Context, repo *repo_model.Repository, doer *user_model.User, branchName, treePath string) error {
+func VerifyBranchProtection(ctx context.Context, repo *repo_model.Repository, doer *user_model.User, branchName string, treePaths []string) error {
 	protectedBranch, err := git_model.GetFirstMatchProtectedBranchRule(ctx, repo.ID, branchName)
 	if err != nil {
 		return err
 	}
 	if protectedBranch != nil {
 		protectedBranch.Repo = repo
-		isUnprotectedFile := false
-		glob := protectedBranch.GetUnprotectedFilePatterns()
-		if len(glob) != 0 {
-			isUnprotectedFile = protectedBranch.IsUnprotectedFile(glob, treePath)
-		}
-		if !protectedBranch.CanUserPush(ctx, doer) && !isUnprotectedFile {
-			return models.ErrUserCannotCommit{
-				UserName: doer.LowerName,
+		for _, treePath := range treePaths {
+			isUnprotectedFile := false
+			glob := protectedBranch.GetUnprotectedFilePatterns()
```

Comment:
It looks like `protectedBranch.GetXxx` (and below) could be moved out the loop. One call one variable is enough?